runtime/drivers/clickhouse/olap.go

```diff
@@ -251,7 +252,86 @@ func (c *connection) InsertTableAsSelect(ctx context.Context, name, sql string,
 		})
 	}
 
-	// merge strategy is also not supported for clickhouse
+	if strategy == drivers.IncrementalStrategyMerge {
+		_, onCluster, err := informationSchema{c: c}.entityType(ctx, "", name)
+		if err != nil {
+			return err
+		}
+		onClusterClause := ""
+		if onCluster {
+			onClusterClause = "ON CLUSTER " + safeSQLName(c.config.Cluster)
+		}
+		// Get the engine info of the given table
+		var engine, engineFull string
+		res, err := c.Execute(ctx, &drivers.Statement{
+			Query:    "SELECT engine, engine_full FROM system.tables WHERE database = currentDatabase() AND name = ?",
+			Args:     []any{name},
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		defer res.Close()
+		if res.Next() {
+			if err := res.Scan(&engine, &engineFull); err != nil {
+				return err
+			}
+		}
+		// Distributed table cannot be altered directly, so we need to alter the underlying table
+		if engine == "Distributed" {
+			// Parse the engine_full string to extract underlying table name
+			// Example engine_full: Distributed('cluster', 'db', 'underlying_table', sharding_key)
+			re := regexp.MustCompile(`Distributed\('[^,]+',\s*'[^']+',\s*'([^']+)'`)
+			matches := re.FindStringSubmatch(engineFull)
+			if len(matches) < 1 {
+				return fmt.Errorf("unable to parse underlying table info from engine expression: %q", engineFull)
+			}
+			name = matches[1]
+		}
+		// create temp table with the same schema
+		tempName := tempName(name)
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("CREATE OR REPLACE TABLE %s %s AS %s", safeSQLName(tempName), onClusterClause, name),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// insert into temp table
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("INSERT INTO %s %s", safeSQLName(tempName), sql),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// list partitions from the temp table using system.parts
+		res, err = c.Execute(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("SELECT DISTINCT partition FROM system.parts WHERE table = '%s'", tempName),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		defer res.Close()
+		// iterate over partitions and replace them in the main table
+		for res.Next() {
+			var part string
+			if err := res.Scan(&part); err != nil {
```

Comment:
I think we should collect all parts first and then loop over collected parts. Though I assume the alter queries are just meta queries but it will be good to free the connection that executed select.